File: solutions/1091_ShortestPathinBinaryMatrix.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var shortestPathBinaryMatrix = function(grid) {
    var length = grid[0].length;
    var height = grid.length;
    //checking edge cases
    if(length < 1 || height  < 1){
        return -1;
    }
    //closed exit
    if(grid[height-1][length-1] === 1){
        return -1;
    }
    //closed entrance
    if(grid[0][0] === 1){
        return -1;
    }
    // 1x1 matrix and open entrance than we need to do 1 step;
    if(length == 1 && height == 1 && grid[0][0] == 0){
        return 1;
    }
    
    
    //Breadth First Search (level by level)
    var queue = [[0,0]] // add first coordinates
    grid[0][0] = 1; //initialize as visited
    
    /*
        0,*,*,*,0,0...
        0,*,E,*,0,0...
        0,*,*,*,0,0,...
    
    */
    var adjacent_coordinates = [[0,1],[1,1],[1,0],[0,-1],[-1,0],[1,-1],[-1,1],[-1,-1]]
    while(queue.length!=0){
        let current_element = queue.shift() // LIFO for bfs;
        let curr_x = current_element[0];
        let curr_y = current_element[1];
        //check all adjacent (примыкающие)
        for(let i=0;i<adjacent_coordinates.length;i++){
            let adjacent = adjacent_coordinates[i]
            let adjacent_x = curr_x + adjacent[0];
            let adjacent_y = curr_y + adjacent[1];
            
            // came to boundaries and we can't exit
            if(adjacent_x < 0 || adjacent_x >= length){
                continue; 
            }
            if(adjacent_y < 0 || adjacent_y >= height){
                continue;
            }
            //if current element is a wall (1) or already visited. we can't go into wall.
            //
            if(grid[adjacent_x][adjacent_y] > 0){
                continue;
            }
            
            //try equal
            if(adjacent_x >= (length - 1) && adjacent_y >= (height - 1) &&  grid[adjacent_x][adjacent_y] == 0){
                return grid[curr_x][curr_y] + 1;
            }
            queue.push([adjacent_x,adjacent_y])
            grid[adjacent_x][adjacent_y] = grid[curr_x][curr_y] + 1;
        }
    }
    return -1 
};

File: solutions/1091_ShortestPathinBinaryMatrix.py
class Solution(object):
    def shortestPathBinaryMatrix(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        length = len(grid[0])
        height = len(grid)
        
        if length < 1 or height < 1:
            return -1
        if grid[height-1][length-1] == 1:
            return -1
        if grid[0][0] == 1:
            return -1
        if length == 1 and height == 1 and grid[0][0]:
            return 1
        
        queue = [(0,0)]
        grid[0][0] = 1
        adjacents_coordinates = [[0,1],[1,1],[1,0],[0,-1],[-1,0],[1,-1],[-1,1],[-1,-1]]
        while queue:
            current = queue.pop(0)
            current_x  = current[0]
            current_y = current[1]
            for i in range(len(adjacents_coordinates)-1):
                adjacents = adjacents_coordinates[i]
                adjacents_x = current_x + adjacents[0]
                adjacents_y = current_y + adjacents[1]
                
                if adjacents_x < 0 or adjacents_x >= length:
                    continue
                
                if adjacents_y < 0 or adjacents_y >= height:
                    continue
                
                if grid[adjacents_x][adjacents_y] > 0:
                    continue
                
                if (adjacents_x >= length - 1 and \
                    adjacents_y >= height-1 and grid[adjacents_x][adjacents_y] == 0):
                    return grid[current_x][current_y]+1
                
                queue.append((adjacents_x,adjacents_y))
                
                grid[adjacents_x][adjacents_y] = grid[current_x][current_y]+1
                
        return -1